google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerCloudMonitoringExporter.java

```diff
@@ -0,0 +1,246 @@
+ * instrumentation scope and upload it via the Google Cloud Monitoring API.
+ */
+class SpannerCloudMonitoringExporter implements MetricExporter {
+
+  private static final Logger logger =
+      Logger.getLogger(SpannerCloudMonitoringExporter.class.getName());
+
+  // This system property can be used to override the monitoring endpoint
+  // to a different environment. It's meant for internal testing only.
+  private static final String MONITORING_ENDPOINT =
+      MoreObjects.firstNonNull(
+          System.getProperty("spanner.test-monitoring-endpoint"),
+          MetricServiceSettings.getDefaultEndpoint());
+
+  // This the quota limit from Cloud Monitoring. More details in
+  // https://cloud.google.com/monitoring/quotas#custom_metrics_quotas.
+  private static final int EXPORT_BATCH_SIZE_LIMIT = 200;
+  private final AtomicBoolean spannerExportFailureLogged = new AtomicBoolean(false);
+  private CompletableResultCode lastExportCode;
+  private final MetricServiceClient client;
+  private final String spannerProjectId;
+  private final AtomicBoolean isShutdown = new AtomicBoolean(false);
+
+  static SpannerCloudMonitoringExporter create(String projectId, @Nullable Credentials credentials)
+      throws IOException {
+    MetricServiceSettings.Builder settingsBuilder = MetricServiceSettings.newBuilder();
+    CredentialsProvider credentialsProvider;
+    if (credentials == null) {
+      credentialsProvider = NoCredentialsProvider.create();
+    } else {
+      credentialsProvider = FixedCredentialsProvider.create(credentials);
+    }
+    settingsBuilder.setCredentialsProvider(credentialsProvider);
+    settingsBuilder.setEndpoint(MONITORING_ENDPOINT);
+
+    org.threeten.bp.Duration timeout = Duration.ofMinutes(1);
+    // TODO: createServiceTimeSeries needs special handling if the request failed. Leaving
+    // it as not retried for now.
+    settingsBuilder.createServiceTimeSeriesSettings().setSimpleTimeoutNoRetries(timeout);
+
+    return new SpannerCloudMonitoringExporter(
+        projectId, MetricServiceClient.create(settingsBuilder.build()));
+  }
+
+  @VisibleForTesting
+  SpannerCloudMonitoringExporter(String projectId, MetricServiceClient client) {
+    this.client = client;
+    this.spannerProjectId = projectId;
+  }
+
+  @Override
+  public CompletableResultCode export(Collection<MetricData> collection) {
+    if (isShutdown.get()) {
+      logger.log(Level.WARNING, "Exporter is shutting down");
+      return CompletableResultCode.ofFailure();
+    }
+
+    lastExportCode = exportSpannerClientMetrics(collection);
+    return lastExportCode;
+  }
+
+  /** Export client built in metrics */
+  private CompletableResultCode exportSpannerClientMetrics(Collection<MetricData> collection) {
+    // Filter spanner metrics
+    List<MetricData> spannerMetricData =
+        collection.stream()
+            .filter(md -> SPANNER_METRICS.contains(md.getName()))
```

Comment:
How does this work with any other MetricExporter that might also be active in the same JVM? This filters the collection so it only exports the metrics that this exporter cares about. But the collection that it was given remains unmodified. Will any other MetricExporter also have to filter out metrics that it does not care about to avoid getting the Spanner-internal metrics? Or are we fine with the Spanner-internal metrics also being exported by other exporters?

Reply:
This exporter will only be added for our Built-in metrics. Customer will use their own exporter and their own set of metrics, so this will not have any impact on other metrics. 

> OpenTelemetrySdk, in which case these metrics won't be created in the global SDK.
Yes you are right , this will not be in the global SDK. 

This check ensures that only our built-in metrics are pushed to Monarch. For example, today all the metrics from the GAX layer are included in our built-in metrics. If the GAX layer introduces a new metric in the future that we don't want to add, this check will reject any additional metrics.

Reply:
Yes, that is true. We will create a new OT object for built in metrics and customer would create their own OT object.